mono/sgen/sgen-gchandles.c

```diff
@@ -432,7 +432,7 @@ scan_for_weak (gpointer hidden, GCHandleType handle_type, int max_generation, gp
 	GCObject *obj = (GCObject *)MONO_GC_REVEAL_POINTER (hidden, is_weak);
 
 	/* If the object is dead we free the gc handle */
-	if (!sgen_is_object_alive (obj))
+	if (!sgen_is_object_alive_for_current_gen (obj))
```

Comment:
Shouldn't we do something similar below, when checking if the object referenced by the field is alive ? If the field_obj is in the major space it will set the field to null.

Also, why don't we scan the weak fields during major collections ?

Reply:
I agree with Vlad here, the second sgen_is_object_alive call must go too.